### PR TITLE
[WIP] Review API for nullability

### PR DESCRIPTION
**Pull request by blairconrad**
Mostly to prompt a discussion. I looked at the public API and tried to see what needed some `?`s.
And where we didn't have `?`s, but added some `Guard.AgainstNull`s. Which should probably be backed by tests.
Some places I just added comments!

### CODE REVIEW

File: src/FakeItEasy/DummyFactory.cs [hunk 1]
```diff
@@ -37,7 +37,8 @@ public object Create(Type type)
         {
             this.AssertThatFakeIsOfCorrectType(type);
 
-            return this.Create();
+            return this.Create(); // this might be null. Same with any implementation of IDummyFactory.
+            // We could allow it and trap in DynamicDummyFactory, returning false there.
```
Comment (thomaslevesque) [inline src/FakeItEasy/DummyFactory.cs]: I don't know... Can a dummy be null?

File: src/FakeItEasy/IArgumentValueFormatter.cs [hunk 2]
```diff
@@ -25,5 +25,6 @@ public interface IArgumentValueFormatter
         /// <param name="argumentValue">The argument value to get as a string.</param>
         /// <returns>A string representation of the value.</returns>
         string GetArgumentValueAsString(object argumentValue);
+        // Check at callsite and throw UsercallbackException if return value is null?
```
Comment (thomaslevesque) [inline src/FakeItEasy/IArgumentValueFormatter.cs]: Sounds good to me

File: src/FakeItEasy/IBootstrapper.cs [hunk 3]
```diff
@@ -24,5 +24,6 @@ public interface IBootstrapper
         /// <returns>A list of absolute paths pointing to assemblies to scan for extension points.</returns>
         [SuppressMessage("Microsoft.Design", "CA1024:UsePropertiesWhereAppropriate", Justification = "A property would not be appropriate here since the operation might perform significant work.")]
         IEnumerable<string> GetAssemblyFileNamesToScanForExtensions();
+        // Consider checking to make sure the return value is not null. Otherwise throw a UserCallbackException.
```
Comment (thomaslevesque) [inline src/FakeItEasy/IBootstrapper.cs]: 👍

File: src/FakeItEasy/Core/DefaultArgumentConstraintManager.cs [hunk 4]
```diff
@@ -17,7 +17,7 @@ public DefaultArgumentConstraintManager(Action<IArgumentConstraint> onConstraint
         public T Matches(Func<T, bool> predicate, Action<IOutputWriter> descriptionWriter)
         {
             this.onConstraintCreated(new MatchesConstraint(predicate, descriptionWriter));
-            return default(T);
+            return default!;
```
[the remark this answers is not in the document]
Comment (blairconrad) [reply src/FakeItEasy/Core/DefaultArgumentConstraintManager.cs]: I know. I really hate it!
Comments galore, if  we do it. I should've commented more as I went.

[the remark this answers is not in the document]
Comment (thomaslevesque) [reply src/FakeItEasy/Core/DefaultArgumentConstraintManager.cs]: Well, that's what the "damn-it operator" is for...

**Comment by thomaslevesque** [comment]
> It's the only `.tt` files that I added `?`s to. T4/compiler bug? If I remove the `.tt` from the project and just dump the source of this file in, the warning disappears.

The fact that it's a T4 template should make no difference whatsoever... The compiler doesn't even look at the .tt files, only the .cs files. I don't understand how it's possible

File: src/FakeItEasy/Creation/IFakeOptions.of.T.cs [hunk 5]
```diff
@@ -32,7 +32,7 @@ public interface IFakeOptions<T> : IHideObjectMembers where T : class
         /// </summary>
         /// <param name="wrappedInstance">The object to delegate calls to.</param>
         /// <returns>Options object.</returns>
-        IFakeOptions<T> Wrapping(T wrappedInstance);
+        IFakeOptions<T> Wrapping(T wrappedInstance); // can we fake an interface and wrap a struct?
```
[the remark this answers is not in the document]
Comment (thomaslevesque) [reply src/FakeItEasy/Creation/IFakeOptions.of.T.cs]: Oh, I see. But it doesn't matter. `T` is the interface, not the struct. If the struct is assignable to `T`, it should work.

File: src/FakeItEasy/DummyFactory.cs (hunk 1 above)
Comment (thomaslevesque) [reply src/FakeItEasy/DummyFactory.cs]: > I generally don't _like_ the idea, but if users think `null` is an appropriate value to return, should we stop them?

I don't think so. There's no technical reason to forbid it. If users want to do that, and it hurts them later... I guess it's their problem?

> If we officially allow null Dummies, should we make sure we don't consider them as usable arguments for constructors where the parameter is a non-nullable reference type?

I'd say no. Non-nullable reference types is a compile time feature. There's no expectation that the constraint should be enforced at runtime. (and even at compile time, it's only a warning, not an error)

Comment (thomaslevesque) [reply src/FakeItEasy/DummyFactory.cs]: > Likewise, if we say that null Dummies are allowed, should our own Dummy resolution algorithm fall back to returning a null when all other creation attempts fail? And if so, should it only do so when trying to generate an argument for a constructor, or should even `A.Dummy<T>` return null?

That's a harder question. I'd say keep the current behavior (i.e. never return a null dummy), but I'm no sure...

File: src/FakeItEasy/Creation/IFakeOptions.of.T.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (blairconrad) [reply src/FakeItEasy/Creation/IFakeOptions.of.T.cs]: How silly of me.

File: src/FakeItEasy/DummyFactory.cs (hunk 1 above)
Comment (blairconrad) [reply src/FakeItEasy/DummyFactory.cs]: > never return a null dummy

… unless a DummyFactory made it?

Yeah, I don't want to start returning nulls willy-nilly. Perhaps we can update the docs (along with whatever code changes to ensure we don't warn on null Dummies coming from DummyFactories) to

> FakeItEasy's own Dummy-creation strategies will never produce a null Dummy, but user-supplied DummyFactories may do so 

And leave it at that.

Comment (thomaslevesque) [reply src/FakeItEasy/DummyFactory.cs]: > … unless a DummyFactory made it?

Exactly

**Comment by blairconrad** [comment]
I'm inclined to add the missing null checks in their own PR. Not looking forward to adding the tests for them, though. We haven't (I think) been putting null guard checks in the specs, as a rule. Do you have a preference for location of the tests?

**Comment by thomaslevesque** [comment]
> I'm inclined to add the missing null checks in their own PR.

👍 

> Not looking forward to adding the tests for them, though. We haven't (I think) been putting null guard checks in the specs, as a rule. Do you have a preference for location of the tests?

Honestly, I'd be OK with *not* writing tests for null checks...

**Comment by blairconrad** [comment]
> Honestly, I'd be OK with not writing tests for null checks...

And here I was, feeling bad because I had that opinion…

**Comment by blairconrad** [comment]
Regarding the "The annotation for nullable reference types should only be used in code within a '#nullable' annotations context." warnings in the T4-generated code, I raised 
[In T4-generated code with nullability on, many "The annotation for nullable reference types should only be used in code within a '#nullable' annotations context." warnings](https://developercommunity.visualstudio.com/content/problem/810959/in-t4-generated-code-with-nullability-on-many-the.html).

**Comment by thomaslevesque** [comment]
> Regarding the "The annotation for nullable reference types should only be used in code within a '#nullable' annotations context." warnings in the T4-generated code, I raised
> [In T4-generated code with nullability on, many "The annotation for nullable reference types should only be used in code within a '#nullable' annotations context." warnings](https://developercommunity.visualstudio.com/content/problem/810959/in-t4-generated-code-with-nullability-on-many-the.html).

I'm really curious to see the explanation for this...

> But we may want to try suppressing with a `#nullable` annotations context if nothing comes of it.

Sure!

File: src/FakeItEasy/DummyFactory.cs [hunk 6]
```diff
@@ -31,20 +31,19 @@ public bool CanCreate(Type type)
...
-            // We could allow it and trap in DynamicDummyFactory, returning false there.
+            return this.Create();
         }
 
         /// <summary>
         /// Creates the dummy.
         /// </summary>
-        /// <returns>The dummy object.</returns>
+        /// <returns>The dummy object. Unlike a dummy provided by built-in FakeItEasy mechanisms, may be <c>null</c>.</returns>
         protected abstract T Create();
```
[the remark this answers is not in the document]
Comment (blairconrad) [reply src/FakeItEasy/DummyFactory.cs]: Or, since I am not a fan of the `null`s, someone who wants this badly enough can implement the interface. Mwu-ha-ha!

[the remark this answers is not in the document]
Comment (thomaslevesque) [reply src/FakeItEasy/DummyFactory.cs]: Fair point! On the other hand, the doc does say the method can return null, so it should probably allow it... Otherwise, we should remove the comment in the doc

[the remark this answers is not in the document]
Comment (blairconrad) [reply src/FakeItEasy/DummyFactory.cs]: Yeah, I was mostly kidding. And of course we'd update the docs to align with code and policy.
You made good points about allowing custom Dummies that are nulls. I think we might as well continue down the path.

[the remark this answers is not in the document]
Comment (blairconrad) [reply src/FakeItEasy/DummyFactory.cs]: Oh. But if the client has the warnings on, and wants to return a null, should their factory instead be a `DummyFactory<ReferenceType?>`? That quells the warning about returning null...

[the remark this answers is not in the document]
Comment (blairconrad) [reply src/FakeItEasy/DummyFactory.cs]: I guess I'm assuming warnings _and_ language features on. Nullability is hard!

[the remark this answers is not in the document]
Comment (blairconrad) [reply src/FakeItEasy/DummyFactory.cs]: I tried adding `[return: MaybeNull]`, but the warning stayed. I'm probably doing something foolish.

**Comment by blairconrad** [comment]
Things are going a little better, but the last two builds failed due to an API difference detected on the Linux build (so far it looks like a line ending problem), but not the Windows (and the API difference reporting failed in the second-last build due to what I think might be the resurgence of an old but in ApprovalTests.Net: https://github.com/approvals/ApprovalTests.Net/issues/159, so I hooted the owners).

I'm surprised all the API approvals aren't failing, if they're not robust against line endings in the files. Investigating.

**Comment by blairconrad** [comment]
@thomaslevesque, at home I have all the nullability warnings removed for the FakeItEasy dll, and it didn't require any additional API changes on top of what we have here, so if you feel like re-reviewing, we can at least know we're happy up to this point.

I'm a little leery of merging now because we'll be introducing hundreds of errors into the build, but if you think we can live with that on on interim basis, we could.
Otherwise, I can tack on my changes for cleaning up "internal nullability warnings" and we can merge in one fell swoop. But we don't have to decide yet.

**Comment by thomaslevesque** [comment]
Maybe we can find a middle ground... Make a long lived `nullability` branch, and merge to that. When it's ready, we'll merge it to master. This way, there won't be a huge PR to review, and we won't merge hundreds of warning to master.

(well, there will be a huge PR when merging nullability to master, but the changes will already have been reviewed)

**Comment by blairconrad** [comment]
And eventually merge the dozens of commits? Realistically, I've made way too many commits for usefulness, I think. It was just convenient to make one per file that I was attempting to clean up, so I had rollbackability.

If I can get `NotNullWhen` picked up by the analyzer shared project soon, I may have a PR that fixes up nullability for all of src in a day or two…

File: src/FakeItEasy/A.cs [hunk 7]
```diff
@@ -80,21 +80,30 @@ public static class A
...
         /// <typeparam name="T">The type of dummy to return.</typeparam>
-        /// <returns>A dummy object of the specified type.</returns>
+        /// <returns>
+        /// A dummy object of the specified type.
+        /// May be null, if a dummy factory is defined that returns null for dummies of type <typeparamref name="T"/>.
+        /// </returns>
         /// <exception cref="ArgumentException">Dummies of the specified type can not be created.</exception>
         [EditorBrowsable(EditorBrowsableState.Advanced)]
         [SuppressMessage("Microsoft.Design", "CA1004:GenericMethodsShouldProvideTypeParameter", Justification = "Used to specify the type of dummy.")]
+        [return: MaybeNull]
```
[the remark this answers is not in the document]
Comment (thomaslevesque) [reply src/FakeItEasy/A.cs]: > Can we live with the `///` comment as above, but strike the attribute? I can.

Sounds good to me

File: src/FakeItEasy/Configuration/OutAndRefParameters.StronglyTyped.cs [hunk 8]
```diff
@@ -6,6 +6,7 @@
 //     the code is regenerated.
 // </auto-generated>
 //------------------------------------------------------------------------------
+#nullable enable
```
[the remark this answers is not in the document]
Comment (thomaslevesque) [reply src/FakeItEasy/Configuration/OutAndRefParameters.StronglyTyped.cs]: Strange... I still don't understand what could be causing the problem in the first place

File: src/FakeItEasy/FakeItEasy.csproj [hunk 9]
```diff
@@ -54,13 +54,13 @@
...
   <PropertyGroup Condition="'$(TargetFramework)' == 'net45'">
-    <DefineConstants>$(DefineConstants);FEATURE_REFLECTION_GETASSEMBLIES;FEATURE_REFLECTIONONLYLOAD;FEATURE_PARAMETERINFO_CUSTOMATTRIBUTES_PROPERTY</DefineConstants>
+    <DefineConstants>$(DefineConstants);FEATURE_REFLECTION_GETASSEMBLIES;FEATURE_REFLECTIONONLYLOAD;FEATURE_PARAMETERINFO_CUSTOMATTRIBUTES_PROPERTY;LACKS_NULLABLE_ATTRIBUTES</DefineConstants>
   </PropertyGroup>
 
   <!-- .NET 4.0 -->
 
   <PropertyGroup Condition="'$(TargetFramework)' == 'net40'">
-    <DefineConstants>$(DefineConstants);FEATURE_REFLECTION_GETASSEMBLIES;FEATURE_REFLECTIONONLYLOAD;FEATURE_EVENT_ARGS_MUST_EXTEND_EVENTARGS</DefineConstants>
+    <DefineConstants>$(DefineConstants);FEATURE_REFLECTION_GETASSEMBLIES;FEATURE_REFLECTIONONLYLOAD;FEATURE_EVENT_ARGS_MUST_EXTEND_EVENTARGS;LACKS_NULLABLE_ATTRIBUTES</DefineConstants>
```
[the remark this answers is not in the document]
Comment (thomaslevesque) [reply src/FakeItEasy/FakeItEasy.csproj]: I don't mind either way... Do as you think is best ;)

[the remark this answers is not in the document]
Comment (blairconrad) [reply src/FakeItEasy/FakeItEasy.csproj]: Then I'll leave it for now. Changing is more work! And this is completely internal. If we decide we hate it next week, there are no repercussions to changing it.
Thanks for indulging me.

**Comment by blairconrad** [comment]
Thanks, @thomaslevesque. I just realized that I should add a test case or two for a null-returning DummyFactory. I can add today or tonight.

And I do have the rest of the warnings removed from `src` in a branch at home. Yay!

File: tests/FakeItEasy.Specs/DummyCreationSpecs.cs [hunk 10]
```diff
@@ -805,6 +834,23 @@ public LazilyCreated()
...
+        }
+
+        public class ClassWhoseDummyIsNullFactory : DummyFactory<ClassWhoseDummyIsNull>
+        {
+            protected override ClassWhoseDummyIsNull Create() => null;
+        }
+
+        public class ClassThatRequiresClassWhoseDummyIsNull
+        {
+            [SuppressMessage("Microsoft.Usage", "CA1801:ReviewUnusedParameters", MessageId = "unused", Justification = "This is just a dummy argument.")]
```
Comment (thomaslevesque) [inline tests/FakeItEasy.Specs/DummyCreationSpecs.cs]: I think we can use `#pragma warning disable CA1801` now. It's fine to leave it as is for now, but we could consider switching to the new syntax in the future. The only benefit of the attribute approach is that it lets us give a justification for the suppression, but we could just do this in a comment.

Comment (blairconrad) [reply tests/FakeItEasy.Specs/DummyCreationSpecs.cs]: You prefer the pragma because it's terser? Any other reason? I assume we'd want to restore after, or are you thinking of a blanket one for the file?

Comment (thomaslevesque) [reply tests/FakeItEasy.Specs/DummyCreationSpecs.cs]: Yes, because it's terser. But now that I think about it, if we restore it immediately afterwards, it makes things even worse... If only we had https://github.com/dotnet/csharplang/issues/1070!

Comment (blairconrad) [reply tests/FakeItEasy.Specs/DummyCreationSpecs.cs]: Yeah, the's the one thing I missed when switching from C++ to C#. (What you describe is very nearly "suppress" from https://docs.microsoft.com/en-us/cpp/preprocessor/warning?view=vs-2019.)
